src/resolve/resolved-varlink.c

```diff
@@ -216,16 +209,53 @@ static void vl_method_resolve_hostname_complete(DnsQuery *query) {
                                                JSON_BUILD_PAIR("family", JSON_BUILD_INTEGER(family)),
                                                JSON_BUILD_PAIR("address", JSON_BUILD_BYTE_ARRAY(p, FAMILY_ADDRESS_SIZE(family)))));
                 if (r < 0)
-                        goto finish;
+                        return r;
 
-                if (!canonical)
-                        canonical = dns_resource_record_ref(rr);
+                if (canonical && !*canonical)
+                        *canonical = dns_resource_record_ref(rr);
 
-                r = json_variant_append_array(&array, entry);
+                r = json_variant_append_array(array, entry);
                 if (r < 0)
-                        goto finish;
+                        return r;
```

Comment:
i'd really update the array first, since this can fail, and if it does then you already updated `*canonical`... if you switch this, then the function will either fail or succeed, but never return half-initialized data...